### DISCUSSION TITLE
v4 HarvestDirectory and preprocessor variable

### DISCUSSION THREAD

**Discussion by Steve0212a** [wixtoolset/issues#7215] [Questions]
I have been unable to get HarvestDirectory working with a preprocessor variable.  I have used wix v3 a lot, but this is my first attempt at v4.

Here is my wixproj (mostly copied from the docs)
```
<Project Sdk="WixToolset.Sdk/4.0.0-rc.2">
  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">
    <CompilerAdditionalOptions>-d SourceFolder=<my path></CompilerAdditionalOptions>
  </PropertyGroup>
  <ItemGroup>
    <HarvestDirectory Include="<path to folder with files>">
      <ComponentGroupName>HarvestedComponents</ComponentGroupName>
      <DirectoryRefId>INSTALLFOLDER</DirectoryRefId>
      <PreprocessorVariable>SourceFolder</PreprocessorVariable>
      <SuppressRootDirectory>true</SuppressRootDirectory>
      <HarvestDirectoryVerboseOutput>true</HarvestDirectoryVerboseOutput>
    </HarvestDirectory>
  </ItemGroup>
  <ItemGroup>
    <PackageReference Include="WixToolset.Heat" Version="4.0.0-rc.2" />
  </ItemGroup>
</Project>
```

However, I am getting this error:
`wix.exe : error WIX0288: The variable 'SourceFolder' with value '<my path>' was previously declared with value '<my path>'.`

In Visual studio in the properties of the wixproj, I have set the Additional build parameters to have the -d... shown above in the wixproj.

Any help would be appreciated.  Thanks in advance

**Comment by barnson** [comment]
Use the MSBuild `/bl` switch to get a binlog and use the log viewer to look for `wix.exe` invocations.

**Comment by Steve0212a** [comment]
So, I am getting different results from MSBuild that I am from Visual Studio, though neither work.  In Visual Studio, I am getting the error in my original post.

I ran MSBuild and it seems to be ignoring my CompilerAdditionalOptions in the wixproj.  Text log attached.  Is there something else I need to add to the msbuild command line to pass that variable to WIX?

[log.txt](https://github.com/wixtoolset/issues/files/10708193/log.txt)

**Reply by robmen** [reply]
To set preprocessor variables use the `DefineConstants` Property instead. Like:
```
   <DefineConstants>SourceFolder=<my path></DefineConstants>
```

**Reply by Steve0212a** [reply]
Updating the wixproj with that solved it in Visual Studio - thanks.

<img width="763" alt="image" src="https://user-images.githubusercontent.com/18215939/218119827-6cc7737e-53bc-442d-8004-07edecaab37a.png">

MSBuild still fails with this error on all the heat generated entries:
`error WIX0150: Undefined preprocessor variable '$(SourceFolder)'`

**Reply by robmen** [reply]
I'd guess that from MSBuild, you're not building the `Debug` and `x64` configuration.

**Comment by Steve0212a** [comment]
I honestly don't use MSBuild that much - usually Visual Studio or dotnet publish.  I just tried this command line and got the same error:

msbuild -r /p:Configuration=DEBUG /p:PlatformTarget=x64 <name>Msi.wixproj

Now that VS is working, I am good, but since v4 is in testing, I was not sure if this was a bug or just my unfamiliarity with MSBUILD...

**Reply by robmen** [reply]
It looks like the name of the platform variable is `Platform` not `PlatformTarget`. You'll want to look at the second line of your .wixproj. IMHO, there is a lot of value in understanding MSBuild. VS and dotnet publish are built on top of it.

**Comment by Steve0212a** [comment]
Ah - got it.  that makes sense.

Thanks for the help!
